glean-core/tests/dual_labeled_counter.rs

```diff
@@ -0,0 +1,527 @@
+    metric.add_sync(&glean, 6);
+    let metric = dual_labeled_counter.get("key3", "category3");
+    metric.add_sync(&glean, 7);
+
+    let snapshot = StorageManager
+        .snapshot_as_json(glean.storage(), "store1", true)
+        .unwrap();
+
+    assert_eq!(
+        json!({
+            "dual_labeled_counter": {
+                "telemetry.dual_labeled_counter": {
+                    "key1": {
+                        "category1": 1,
+                        "category2": 2,
+                        "__other__": 6
+                    },
+                    "key2": {
+                        "category1": 3,
+                        "category2": 4
+                    },
+                    "__other__": {
+                        "category1": 5,
+                        "__other__": 7
+                    }
+                }
+            }
+        }),
+        snapshot
+    );
+}
+
+#[test]
+fn dynamic_labels_too_long() {
+    let (glean, _t) = new_glean(None);
+    let dual_labeled_counter = DualLabeledCounterMetric::new(
+        CommonMetricData {
+            name: "dual_labeled_counter".into(),
+            category: "telemetry".into(),
+            send_in_pings: vec!["store1".into()],
+            disabled: false,
+            lifetime: Lifetime::Ping,
+            ..Default::default()
+        },
+        None,
+        None,
+    );
+
+    let metric = dual_labeled_counter.get("1".repeat(112), "2".repeat(112));
+    metric.add_sync(&glean, 1);
+
+    let snapshot = StorageManager
+        .snapshot_as_json(glean.storage(), "store1", true)
+        .unwrap();
+
+    // We end up with 2x InvalidLabel errors, one each for key and category which are both too long.
+    assert_eq!(
+        json!({
+            "dual_labeled_counter": {
+                "telemetry.dual_labeled_counter": {
+                    "__other__": {
+                        "__other__": 1
+                    }
+                }
+            },
+            "labeled_counter": {
+                "glean.error.invalid_label": { "telemetry.dual_labeled_counter": 2 }
+            }
+        }),
+        snapshot
+    );
+}
+
+#[test]
+fn dynamic_labels_regex_mismatch() {
+    let (glean, _t) = new_glean(None);
+    let dual_labeled_counter = DualLabeledCounterMetric::new(
+        CommonMetricData {
+            name: "dual_labeled_counter".into(),
+            category: "telemetry".into(),
+            send_in_pings: vec!["store1".into()],
+            disabled: false,
+            lifetime: Lifetime::Ping,
+            ..Default::default()
+        },
+        None,
+        None,
+    );
+
+    let metric = dual_labeled_counter.get("non-ASCII-key�", "non-ASCII-cat�");
+    metric.add_sync(&glean, 1);
+
+    let snapshot = StorageManager
+        .snapshot_as_json(glean.storage(), "store1", true)
+        .unwrap();
+
+    // We end up with 2x InvalidLabel errors, one each for key and category which both contain
+    // invalid characters.
```

Comment:
Why? I thought we allow any UTF-8 characters for dynamic labels?

Reply:
Ah, the `validate_dynamic_key_and_or_category` -> `label_is_valid` call chain has the `is_ascii` check. That should probably go away.

Reply:
Once that is gone we should also have an explict test that uses a label that has `\x1e` in there. We don't want _those_ labels (and I guess we should call that out in the docs?)